File: Editorial/3546_Equal_Sum_Grid_Partition_I/C/2_Matrix_Rotation_Row_wise_Enumeration.c
int** rotation(int** grid, int gridSize, int* gridColSize, int* returnSize,
               int** returnColumnSizes) {
    int m = gridSize, n = gridColSize[0];
    int** tmp = (int**)malloc(sizeof(int*) * n);
    for (int i = 0; i < n; i++) {
        tmp[i] = (int*)malloc(sizeof(int) * m);
    }
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            tmp[j][m - 1 - i] = grid[i][j];
        }
    }
    *returnSize = n;
    *returnColumnSizes = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) {
        (*returnColumnSizes)[i] = m;
    }
    return tmp;
}

bool canPartitionGrid(int** grid, int gridSize, int* gridColSize) {
    long long total = 0;
    int m = gridSize, n = gridColSize[0];
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            total += grid[i][j];
        }
    }
    for (int k = 0; k < 2; k++) {
        long long sum = 0;
        m = gridSize;
        n = gridColSize[0];
        for (int i = 0; i < m - 1; i++) {
            for (int j = 0; j < n; j++) {
                sum += grid[i][j];
            }
            if (sum * 2 == total) {
                return true;
            }
        }
        int returnSize, *returnColumnSizes;
        grid = rotation(grid, gridSize, gridColSize, &returnSize,
                        &returnColumnSizes);
        gridSize = returnSize;
        gridColSize = returnColumnSizes;
    }
    return false;
}

File: Editorial/3501-3600/3546_Equal_Sum_Grid_Partition_I/JavaScript/2_Matrix_Rotation_Row_wise_Enumeration.js
var canPartitionGrid = function (grid) {
    let total = 0;
    let m = grid.length;
    let n = grid[0].length;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            total += grid[i][j];
        }
    }
    for (let k = 0; k < 2; k++) {
        let sum = 0;
        m = grid.length;
        n = grid[0].length;
        for (let i = 0; i < m - 1; i++) {
            for (let j = 0; j < n; j++) {
                sum += grid[i][j];
            }
            if (sum * 2 === total) {
                return true;
            }
        }
        grid = rotation(grid);
    }
    return false;
};

function rotation(grid) {
    const m = grid.length,
        n = grid[0].length;
    const tmp = Array.from({ length: n }, () => Array(m).fill(0));
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            tmp[j][m - 1 - i] = grid[i][j];
        }
    }
    return tmp;
}
